src/Analyzers/CSharp/Analyzers/UseCompoundAssignment/CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer.cs

```diff
@@ -84,5 +99,108 @@ private void AnalyzeCoalesceExpression(SyntaxNodeAnalysisContext context)
+
+            return whenTrue != null && assignment != null;
+        }
+
+        private void AnalyzeIfStatement(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = context.SemanticModel;
+
+            var ifStatement = (IfStatementSyntax)context.Node;
+
+            var option = context.GetAnalyzerOptions().PreferCompoundAssignment;
+
+            // Bail immediately if the user has disabled this feature.
+            if (!option.Value)
+                return;
+
+            if (ifStatement.Else != null)
+                return;
+
+            if (!GetWhenTrueAssignment(ifStatement, out _, out var assignment))
+                return;
+
+            if (!IsReferenceEqualsNullCheck(semanticModel, ifStatement.Condition, cancellationToken, out var testedExpression))
+                return;
+
+            // have    if (x == null) x = y;
+            // ensure that 'x' in both locations are suitably equivalent.
+            var syntaxFacts = CSharpSyntaxFacts.Instance;
+            if (!syntaxFacts.AreEquivalent(testedExpression, assignment.Left))
+                return;
+
+            // Syntactically looks promising.  But we can only safely do this if 'expr'
+            // is side-effect-free since we will be changing the number of times it is
+            // executed from twice to once.
+            if (!UseCompoundAssignmentUtilities.IsSideEffectFree(
+                    syntaxFacts, testedExpression, semanticModel, cancellationToken))
+            {
+                return;
+            }
+
+            // Good match.
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                ifStatement.IfKeyword.GetLocation(),
+                option.Notification.Severity,
+                additionalLocations: ImmutableArray.Create(ifStatement.GetLocation()),
+                properties: null));
+        }
+
+        private bool IsReferenceEqualsNullCheck(
+            SemanticModel semanticModel,
+            ExpressionSyntax condition,
+            CancellationToken cancellationToken,
+            [NotNullWhen(true)] out ExpressionSyntax? testedExpression)
+        {
+            testedExpression = null;
+            if (condition is BinaryExpressionSyntax(SyntaxKind.EqualsExpression) { Right: LiteralExpressionSyntax(SyntaxKind.NullLiteralExpression) } binaryExpression)
+            {
+                // Ensure that if we are using `==` that it's not an overloaded operator.  One known exception is
+                // System.String.  Even though `==` is overloaded, it has the same semantics as ReferenceEquals(null) so
+                // it's safe to convert.
+                var symbol = semanticModel.GetSymbolInfo(binaryExpression, cancellationToken).Symbol;
+                if (symbol is null || !symbol.IsUserDefinedOperator() || symbol.ContainingType.SpecialType == SpecialType.System_String)
```

Comment:
null check looks unnecessary? The extension method will return `false` if symbol is null:

https://github.com/dotnet/roslyn/blob/373949ea39831d47005b2868eb9fee175e890490/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs#L234-L235

Reply:
it's not an `&&` it's an `||`.

Reply:
fair point :)  i'm ok either way.